kroxylicious-filters/kroxylicious-record-validation/src/main/java/io/kroxylicious/proxy/filter/schema/validation/bytebuf/JsonSchemaBytebufValidator.java

```diff
@@ -12,31 +12,90 @@
     private final JsonValidator jsonValidator;
+    private final Long globalId;
+    private final HeadersHandler keyHeaderHandler;
+    private final HeadersHandler valueHeaderHandler;
+    private final IdHandler keyIdHandler;
+    private final IdHandler valueIdHandler;
 
     public JsonSchemaBytebufValidator(Map<String, Object> schemaResolverConfig, Long globalId) {
+        this.globalId = globalId;
         this.jsonValidator = new JsonValidator(schemaResolverConfig, Optional.of(ArtifactReference.fromGlobalId(globalId)));
+        this.keyHeaderHandler = buildHeaderHandler(true);
+        this.keyIdHandler = buildIdHandler(true);
+
+        this.valueHeaderHandler = buildHeaderHandler(false);
+        this.valueIdHandler = buildIdHandler(false);
     }
 
     @Override
     public CompletionStage<Result> validate(ByteBuffer buffer, Record record, boolean isKey) {
         try {
-            JsonValidationResult jsonValidationResult = jsonValidator.validateByArtifactReference(buffer.clear());
+            // If the record includes a schema id, validate that it is consistent with the expected globalId.
+            Optional<Long> extractedGlobalId = extractGlobalIdFromRecord(buffer, record, isKey);
+            if (extractedGlobalId.filter(e -> !e.equals(globalId)).isPresent()) {
+                return CompletableFuture
+                        .completedStage(new Result(false, "Unexpected schema id in record (%d), expecting %d".formatted(extractedGlobalId.get(), globalId)));
+            }
+
+            JsonValidationResult jsonValidationResult = jsonValidator.validateByArtifactReference(buffer);
             return jsonValidationResult.success() ? Result.VALID_RESULT_STAGE
                     : CompletableFuture.completedFuture(new Result(false, jsonValidationResult.toString()));
         }
         catch (RuntimeException ex) {
-            return CompletableFuture.completedFuture(new Result(false, ex.getMessage()));
+            return CompletableFuture.completedStage(new Result(false, ex.getMessage()));
+        }
+    }
+
+    private Optional<Long> extractGlobalIdFromRecord(ByteBuffer buffer, Record kafkaRecord, boolean isKey) {
+        if (kafkaRecord.headers().length > 0) {
+            var recordHeaders = new RecordHeaders(kafkaRecord.headers());
+            var headerHandler = isKey ? keyHeaderHandler : valueHeaderHandler;
+            var ref = headerHandler.readHeaders(recordHeaders);
+
+            if (ref.getGlobalId() != null) {
+                return Optional.of(ref.getGlobalId());
+            }
+        }
+
+        var idHandler = isKey ? keyIdHandler : valueIdHandler;
+        var minBytes = 1 + idHandler.idSize();
+        if (buffer.remaining() > minBytes && buffer.get(buffer.position()) == AbstractKafkaSerDe.MAGIC_BYTE) {
+            buffer.get(); // ignore magic
+            return Optional.of(idHandler.readId(buffer).getGlobalId());
         }
+        return Optional.empty();
```

Comment:
Just checking the intent here, the filter also validates JSON messages with no global id in the message or header. 

We should make sure to trumpet this, it's a good feature that we can JSON schema check JSON blobs from an arbitrary client that doesn't integrate with schema registries.

Reply:
Yes, that's the intent.  I'll make sure the docs cover that.